### PR TITLE
Encode trait impls using nested Tables

### PR DESCRIPTION
**Pull request by Aaron1011**
We now decode trait impls DefIds only as needed.

### CODE REVIEW

**Comment by petrochenkov** [comment]
This is relevant to perf regressions in https://github.com/rust-lang/rust/pull/74682.

**Comment by petrochenkov** [comment]
It would be great if @eddyb could review this.

**Comment by Aaron1011** [comment]
> This is relevant to perf regressions in #74682.

Yeah, my hope is to at least partially mitigate the impact of adding additional crates to the sysroot.

**Comment by Dylan-DPC** [comment]
@Aaron1011 any updates on this?

**Comment by petrochenkov** [comment]
r? @eddyb 
eddyb is also pinged in a zulip PM.

File: src/librustc_metadata/rmeta/mod.rs [hunk 1]
```diff
@@ -286,6 +280,7 @@ define_tables! {
     mir: Table<DefIndex, Lazy!(mir::Body<'tcx>)>,
     promoted_mir: Table<DefIndex, Lazy!(IndexVec<mir::Promoted, mir::Body<'tcx>>)>,
     unused_generic_params: Table<DefIndex, Lazy<FiniteBitSet<u64>>>,
+    trait_impls: Table<u32, Lazy!(Table<DefIndex, Lazy!([DefIndex])>)>,
```
Comment (eddyb) [inline src/librustc_metadata/rmeta/mod.rs:283]: We don't have sparse tables yet (I've been meaning to do that but haven't gotten around to it) so you should look at how much this change increases the size of `libcore-*.rlib` and `libstd-*.rlib` etc.

File: src/librustc_metadata/rmeta/table.rs [hunk 2]
```diff
@@ -93,24 +93,34 @@ impl<T: Encodable> FixedSizeEncoding for Option<Lazy<T>> {
...
+            let len: u32 = len.try_into().unwrap();
+
+            len.write_to_bytes(&mut b[u32::BYTE_LEN..]);
+        }
+    };
+}
 
-        let len = self.map_or(0, |lazy| lazy.meta);
-        let len: u32 = len.try_into().unwrap();
+impl<I: Idx, T: Encodable> FixedSizeEncoding for Option<Lazy<Table<I, Lazy<[T], usize>>, usize>> {
```
Comment (eddyb) [inline src/librustc_metadata/rmeta/table.rs:118]: `Lazy<[T], usize>` is unnecessary specific here.

File: src/librustc_metadata/rmeta/table.rs [hunk 3]
```diff
@@ -93,24 +93,34 @@ impl<T: Encodable> FixedSizeEncoding for Option<Lazy<T>> {
     }
 }
 
-impl<T: Encodable> FixedSizeEncoding for Option<Lazy<[T]>> {
-    fixed_size_encoding_byte_len_and_defaults!(u32::BYTE_LEN * 2);
+macro_rules! meta_body {
```
Comment (eddyb) [inline src/librustc_metadata/rmeta/table.rs:96]: `impl_fixed_size_encoding_for_usize_meta` would be a better name, but also, have you tried this?
```diff
-impl<T: Encodable> FixedSizeEncoding for Option<Lazy<[T]>> {
+impl<T: ?Sized + Encodable + LazyMeta<Meta = usize>> FixedSizeEncoding for Option<Lazy<T>> {
```
Although perhaps that is still considered overlapping, I'm not sure of the coherence rules around this.

File: src/librustc_metadata/rmeta/mod.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_metadata/rmeta/mod.rs:283]: You could maybe make this cheaper by removing the inner table and keeping a cache that's keyed on the `u32` with a value of `FxHashMap<DefIndex, Lazy<[DefIndex]>>` for every crate.

Also, the `u32`, should that be `CrateNum`? That would be clearer and as a `Table` it shouldn't cause problems I don't think.

**Review by eddyb — COMMENTED** [review]
I've left some comments but I don't think there's a point in doing this if it's not a measurable perf win.

Also I keep forgetting about this, but perf.r-l.o should show stdlib crate sizes (and/or the size of the `lib.rmeta` file inside an `rlib`), in case something makes them increase substantially. cc @Mark-Simulacrum

(Or maybe we can have some other way to get that information? It can be obtained for arbitrary builds by downloading the `rust-std` component)

**Comment by eddyb** [comment]
Oh, something else: as per #75008, it might make more sense to index the impls by the simplified type, not just the trait. Not sure what the best way to do that is, though.

**Comment by Aaron1011** [comment]
Closing this for now - I may revisit it if I come up with a way to make it a noticable perf win. However, it may be that this never really matters in practice (e.g. most crates don't have very many trait impls, or access to them is usually all-or-nothing).

**Pull request closed without merging**